test/jdk/java/nio/channels/FileChannel/TransferFromExtend.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8303260
+ * @summary Test transferFrom to a position greater than size
+ * @library .. /test/lib
+ * @build jdk.test.lib.RandomFactory
+ * @run junit TransferFromExtend
+ * @key randomness
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+
+import static java.nio.file.StandardOpenOption.*;
+
+import jdk.test.lib.RandomFactory;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class TransferFromExtend {
+    private static final Random RND = RandomFactory.getRandom();
+
+    private static final Path DIR = Path.of(System.getProperty("test.dir", "."));
+
+    private static Stream<Arguments> paramProvider(int transferSizeMin,
+                                                   int transferSizeMax) {
+        List<Arguments> list = new ArrayList<Arguments>();
+        int sizeDelta = transferSizeMax - transferSizeMin;
+        for (int i = 0; i < 10; i++) {
+            Arguments args =
+                Arguments.of(RND.nextInt(1024),
+                             transferSizeMin + RND.nextInt(sizeDelta),
+                             1 + RND.nextInt(2047));
+            list.add(args);
+        }
+        return list.stream();
+    }
+
+    //
+    // transfer size must be greater than the threshold
+    // sun.nio.ch.FileChannelImpl::MAPPED_TRANSFER_THRESHOLD (16K)
+    // for a mapped transfer to be used when direct is unavailable
+    //
+    private static Stream<Arguments> fastParamProvider() {
+        return paramProvider(16*1024 + 1, 500*1024);
+    }
+
+    private static Stream<Arguments> readingByteChannelParamProvider() {
+        return paramProvider(1, 64*1024);
+    }
+
+    //
+    // Test the direct and memory-mapped paths. At present the direct path
+    // is implemented only on Linux. The mapped path will be taken only if
+    // there is no direct path and the size of the transfer is more than
+    // sun.nio.ch.FileChannelImpl::MAPPED_TRANSFER_THRESHOLD (16K).
+    // This method therefore tests the direct path on Linux and the mapped
+    // path on other platforms.
+    //
```

Comment:
Would be possible to change the method description to /* .. */. Also I wonder if we can trim down this comment as it will go stale very quickly due to churn in the implementation. Maybe reduce it down to say that it tests transfer from a channel type that may be optimized?